### ISSUE TITLE
Discuss adding --clean-obo option to convert

### ISSUE THREAD

**Issue by beckyjackson** [ontodev/robot#995]
Some ontologies contain axioms that will not properly convert to OBO, and the OBO writer sometimes fails with no good information on how to fix the issue - see https://github.com/OBOFoundry/COB/issues/168

We should consider adding a hack to `convert` that removes these bad axioms (e.g., `inverse` in subClass axioms). My main concern with this is how to handle is multiple annotations (label, definition, comment). From above discussion:
> Do we drop all of them, or do we just pick a random one to drop? That could result in some faulty outputs and spurious diffs.

cc @matentzn

**Comment by jamesaoverton** [comment]
I agree that this is a good feature to add. It needs to be deterministic, and @matentzn's suggestion of first in lexical order, and INFO for dropped annotations, sounds good to me.

Are you volunteering to implement, or find someone to implement this @matentzn?

**Comment by matentzn** [comment]
I will send my sensors out. :)

**Comment by gouttegd** [comment]
Should that `--clean-obo` option also direct the removal of the “untranslated axioms” (which normally end up in the `owl-axioms` header tag)?

I would argue that it should, given the number of pipelines in the wild that contain variations of:

```sh
robot convert -i input.owl -f obo -o tmp.obo && grep -v ^owl-axioms tmp.obo > output.obo
```

Those untranslated axioms are clearly not wanted by a lot of people! And it’s a pretty low hanging fruit, since the OWLAPI already has an option to prevent them from being included in the output, all ROBOT has to do is to use that option.

**Comment by matentzn** [comment]
Yes! The logger should probably Report which axioms are dropped by such an option, either with info or warning.

**Comment by gouttegd** [comment]
OK, then should it also drop GCI axioms? This is [something that has been requested for CL](https://github.com/obophenotype/cell-ontology/issues/2856), on the rationale that many OBO parsers are too simplistic, do not expect the presence of GCIs, and therefore do not interpret them correctly.

I am inclined to say it should (or at least, that it should _allow_ to do it). If we add a “clean OBO“ output option, it should be suitable for existing use cases.

**Comment by matentzn** [comment]
Formally ideally, there should be sepearte options for all subproblems

- "drop GCIs" fits best under `--axioms gci`
- `--drop-duplicate-labels`
- `--drop-duplicate-comments`
- `--axioms obo-format-incompatible`

And whatever else needs to go there, and then have a super option `--clean-obo` that sets the truth values of all of the above to `true`?

**Comment by gouttegd** [comment]
I don’t really like the idea of cluttering the `convert` command with a bunch of OBO-specific options. It would be better then to have a dedicated `obo-export` command, which could take as many options as needed.

Rather, I was imaging that the single `--clean-obo` option could accept, in addition to a single true or false parameter, a comma-separated list of keywords that dictate the exact behaviour(s) to implement:

* `strip-untranslated-axioms`
* `strip-gci-axioms`
* `fix-supernumerary-annotations`

Examples:

* `--clean-obo false`: produce “normal” OBO output (default behaviour);
* `--clean-obo strip-untranslated-axioms,strip-gci-axioms`: remove all untranslated axioms as well as GCI axioms;
* `--clean-obo true`: clean everything we can (equivalent to `--clean-obo strip-untranslated-axioms,strip-gci-axioms,fix-supernumerary-annotations`).

**Comment by matentzn** [comment]
A sorry, I was talking about extending `remove`, not `convert`; 

If you think extending convert is better, your suggestion is fine, bit you should use `--the "usual way of grouping lists"` in ROBOT rather than using `,` I think!

**Comment by gouttegd** [comment]
`remove` could probably be extended to add at least a `--axioms gci` axioms, but I am not keen on a `--axioms obo-format-incompatible` option.

The easiest way to drop OBO incompatible axioms is to rely on the built-in feature from the OWLAPI:

```java
OWLAPIOwl2Obo oboConverter = new OWLAPIOwl2Obo(ontologyManager);
oboConverter.setDiscardUntranslatable(true);
OBODoc oboDoc = oboConverter.convert(ontology);
// Send the oboDoc to a OBOFormatWriter for actual writing
```

But this can only be used in the context of a writing operation, just before we actually write the output.

Trying to remove OBO incompatible axioms outside of such of context (e.g. during a `remove` operation) would require ROBOT to implement its own logic to figure out which axioms are “OBO incompatible” and must be removed. This is I believe a needless complication, and one that carries the risk that the OWLAPI and ROBOT could have a different idea of which axioms are “untranslatable”.

**Comment by gouttegd** [comment]
Regarding the treatment of “supernumerary annotations”, I propose the following behaviour:

**For labels**

* If one label is language-agnostic and the others have a language tag, we keep only the language-agnostic one.
* Otherwise, we do as already proposed earlier: sort the labels lexicographically and keep whichever label comes out first.

**For definitions**

Same treatment as the labels.

**For comments**

* If one comment is language-agnostic and the others have a language tag, we keep only the language-agnostic one.
* Otherwise, we sort the comments lexicographically and _merge_ them into a single comment (adding a space between them).

Rationale for merging the comments: (1) Because that is what has been requested for CL. (2) It avoids a loss of potentially useful information. (3) Merging comments should be relatively harmless, contrary to merging labels (which would be a very bad idea, as it is unlikely that a label formed by merging several labels together could end up being meaningful; likewise for definitions).

**Comment by matentzn** [comment]
> Trying to remove OBO incompatible axioms outside of such of context (e.g. during a remove operation) would require ROBOT to implement its own logic to figure out which axioms are “OBO incompatible” and must be removed. This is I believe a needless complication, and one that carries the risk that the OWLAPI and ROBOT could have a different idea of which axioms are “untranslatable”.

Yes, I am convinced now you are right.

> If one label is language-agnostic and the others have a language tag, we keep only the language-agnostic one.

Yes

> Otherwise, we do as already proposed earlier: sort the labels lexicographically and keep whichever label comes out first.

Yes

> merging comments

I wouldn't bother with language tags. I would 

1. Sort comments and pick first the same way as done with the above unless
2. a new option `--merge-comments` is set to "true" (I don't care if this is set to true by default; but I do think this behaviour should be overwriteable)

I still think the problem of `supernumerary annotations` might be useful in other contexts, not just OBO format, but we can repurpose the code later. The OBO Format use case is by far the dominant one for now.

**Comment by matentzn** [comment]
All sounds great now; sold!

**Comment by gouttegd** [comment]
For convenience, I would additionally propose two more keywords parameters to the `--clean-obo` option:

* `--clean-obo strict`: do only what is necessary to produce a valid OBO file, without removing anything else. This would be equivalent to `--clean-obo "drop-extra-labels drop-extra-definitions drop-extra-comments"`, since a file containing more than one label, definition, or comment per term is _not_ a valid OBO file. “Untranslatable“ axioms would _not_ be stripped, since a OBO file containing a `owl-axioms:` header tag with untranslatable axioms is still a perfectly valid OBO file, even though most people don’t want those axioms in their files. Likewise for GCI axioms, they may not be wanted but they don’t make a file invalid.
* `--clean-obo simple`: do what is necessary to produce a valid OBO file, _and_ also remove the (valid) stuff that most people do not want (untranslatable axioms and GCI axioms). Equivalent to `--clean-obo "drop-untranslatable drop-gci drop-extra-labels drop-extra-definitions drop-extra-comments"`. This would therefore produce a “simple” OBO file, suitable for people who use non-conformant OBO parsers.

That way, most people would likely never need to list the options they need in full details -- `strict` and `simple` would already cover the majority of use cases.

`--clean-obo true` could be equivalent to either `--clean-obo strict` or `--clean-obo simple`.

**Comment by matentzn** [comment]
I like it!

**Comment by jamesaoverton** [comment]
Excellent discussion, and much appreciated!
